### DISCUSSION TITLE
GraphQL batch update

### DISCUSSION THREAD

**Discussion by amaster507** [geldata/gel#3467] [Ideas]
I am liking the generated GraphQL API so far for basic CRUD operations. One thing I requested for a long time from Dgraph could also be of value here, and that is batch GraphQL update mutations.

Let's say I have a set of filters and for each filter I want to apply a different data update. This might look like this right now:

```GraphQL
mutation {
  update1: update_Setting(filter: { id: { eq: "123" } } data: { value: { set: "foo" } }) { id value }
  update2: update_Setting(filter: { id: { eq: "456" } } data: { value: { set: "bar" } }) { id value }
}
```
While this works this becomes harder for FE application developers because they have to either make multiple mutation calls (which defeats the purpose of GraphQL) or they have to dynamically generate the multi-aliased single mutation operation with separate GraphQL variables for each update block separately.

If there was a way to do something like:
```GraphQL
mutation ($batchUpdate: [UpdateOrganizationBatch!]!) {
  batchUpdate_Setting(batch: $batchUpdate) { id value }
}
```
```JSON
{
  "variables": {
    "batchUpdate": [
      { "filter": { "id": { "eq": "123" } }, "data": { "value": { "set": "foo" } } },
      { "filter": { "id": { "eq": "456" } }, "data": { "value": { "set": "bar" } } }
    ]
  }
}
```

This would allow the response to be easier to work as well because it would be like:
```JSON
{
  "data": {
    "batchUpdate_Setting": [
      { "id": "123", "value": "foo" },
      { "id": "456", "value": "bar" },
    ]
  }
}
```
instead of a separate aliased update as is currently the only possible solution beside multiple round trips:
```JSON
{
  "data": {
    "update1": [
      { "id": "123", "value": "foo" }
    ],
    "update2": [
      { "id": "456", "value": "bar" }
    ]
  }
}
```

---

All of this could be a little simpler and was an easier suggestion on Dgraph because the update mutation was generated like:
```GraphQL
mutation {
  updateOrganization(input: UpdateOrganization!): OrganizationUpdatePayload
}
input UpdateOrganization {
  filter: UpdateFilter!
  set: UpdateRef
}
```
So there the change would be easier backwards compatible by just changing the input to be instead `input: [UpdateOrganization!]!` and the rest would work the same. But in edgeDB it would not be an easy backwards compatible change to extend the existing generated update mutations to work with batches.

And the discussion around this in that other generated GraphQL layer from Dgraph can be found: https://discuss.dgraph.io/t/cannot-update-multiple-nodes-at-once/12188?u=amaster507

**Comment by elprans** [comment]
Interesting idea.  This is conceptually similar to client-level batching of EdgeQL queries (a-la [`executemany() in asyncpg`](https://magicstack.github.io/asyncpg/current/api/index.html#asyncpg.connection.Connection.executemany)).  The GraphQL implementation of course wouldn't need any of that and the GraphQL -> EdgeQL translator simply needs to transform `batchUpdate_Foo` into a `FOR` loop.  @vpetrovykh, thoughts?
